flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateGlobalWindowFunction.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.aggregate
+
+import org.apache.flink.streaming.api.functions.windowing.RichAllWindowFunction
+import org.apache.flink.types.Row
+import org.apache.flink.util.Collector
+import org.apache.flink.streaming.api.windowing.windows.Window
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.table.functions.Accumulator
+
+import java.lang.Iterable
+import org.apache.flink.table.functions.AggregateFunction
+
+/**
+  * Computes the final aggregate value from incrementally computed aggreagtes.
+  *
+  * @param aggregates The aggregates to be computed
+  * @param aggFields the fields on which to apply the aggregate.
+  * @param forwardedFieldCount The fields to be carried from current row.
+  */
+class DataStreamProcTimeAggregateGlobalWindowFunction[W <: Window](
```

Comment:
A `ProcessFunction` would look like this:

- `open()` In open we create two types of state: 1) the accumulators, which is a `ValueState[Row]`, 2) the rows we need to retract and their processing timestamp. This can be a `ListState[Tuple2[Long, Row]]`.
- `processElement()` is called for each arriving row. We accumulate the row to the accumulator (fetched from the state) and retract all value from the list which fall out of the range (could also be none). We get the current processing time from the Context object. Since the list is sorted on insertion order, we take the head elements. We emit the row with the updated accumulator and append it to the list of rows to retract together with its processing timestamp (again, current time taken from the Context object). Finally we update both states.
- `onTimer()` is not required for processing time OVER window.

If we want to make it more efficient, we could just put those fields into the retraction rows which are actually used by the aggregation function and not the complete row.

What do you think?